front/ui/ui-charts/src/trackOccupancyDiagram/components/layers/OccupancyZonesLayer.tsx

```diff
@@ -17,19 +54,168 @@ const OccupancyZonesLayer = ({
+        }
+
+        // if so and it's an even index, move it to the bottom, if it's an odd index, move it to the top
+        else if (zoneCounter < MAX_ZONES) {
+          if (arrivalTime >= primaryArrivalTime) {
+            if (zoneCounter % 2 === 0) {
+              yPosition -= yOffset;
+            } else {
+              yPosition += yOffset;
+            }
+          }
+
+          // update the last departure time if the current zone is longer
+          if (departureTime >= lastDepartureTime) lastDepartureTime = departureTime;
+
+          instructions.push({
+            type: 'occupancyZone',
+            zone,
+            offsetY: trackY + yPosition,
+            isSelected: zone.trainId === selectedTrainId,
+          });
+
+          zoneCounter++;
+          yOffset += Y_OFFSET_INCREMENT;
+          zoneIndex++;
+        }
+
+        // else, if there are too much trains:
+        else {
+          const nextIndex = filteredOccupancyZones.findIndex(
+            (filteredZone, i) => i > zoneIndex && filteredZone.arrivalTime >= lastDepartureTime
+          );
+
+          const remainingTrainsNb = nextIndex - zoneIndex;
+
+          instructions.push({
+            type: 'remainingTrains',
+            amount: remainingTrainsNb,
+            time: (zone.arrivalTime + zone.departureTime) / 2,
+            offsetY: trackY,
+          });
+
+          zoneIndex += remainingTrainsNb;
+        }
+      }
+    });
+
+    return sortBy(instructions, (instruction) =>
+      instruction.type === 'occupancyZone' && instruction.isSelected ? 0 : 1
+    );
+  }, [occupancyZones, selectedTrainId, topPadding, tracks]);
+
   const drawingFunction = useCallback<DrawingFunction>(
     (ctx, stcContext) => {
-      drawOccupancyZones(ctx, stcContext, {
-        tracks,
-        occupancyZones,
-        selectedTrainId,
-        position,
-        topPadding,
+      zonesToDraw.forEach((instruction) => {
+        switch (instruction.type) {
+          case 'occupancyZone':
+            drawOccupationZone(ctx, stcContext, {
+              zone: instruction.zone,
+              position,
+              yOffset: instruction.offsetY,
+              isSelected: instruction.isSelected,
+            });
+            break;
+          case 'remainingTrains':
+            drawRemainingTrainsBox(ctx, stcContext, {
+              position,
+              time: instruction.time,
+              yOffset: instruction.offsetY,
+              remainingTrainsNb: instruction.amount,
+            });
+            break;
+        }
+      });
+    },
+    [position, zonesToDraw]
+  );
+
+  const pickingFunction = useCallback<PickingDrawingFunction>(
+    (imageData, { registerPickingElement, getTimePixel, getSpacePixel }, scalingRatio) => {
+      zonesToDraw.forEach((instruction) => {
+        if (instruction.type === 'occupancyZone') {
+          const x = getTimePixel(instruction.zone.arrivalTime);
+          const y = instruction.offsetY + getSpacePixel(position);
```

Comment:
getSpacePixel(position) corresponds to the global y offset right ? Maybe you can declare it outside the forEach loop